provider/query_parameter.go

```diff
@@ -0,0 +1,58 @@
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+// QueryParameter represents an HTTP query parameter specified for use with
+// a given map instance.
+type QueryParameter struct {
+	Name  string `toml:"name"`
+	Token string `toml:"token"`
+	Type  string `toml:"type"`
+	SQL   string `toml:"sql"`
+	// DefaultSQL replaces SQL if param wasn't passed. Either default_sql or
+	// default_value can be specified
+	DefaultSQL   string `toml:"default_sql"`
+	DefaultValue string `toml:"default_value"`
+}
+
+// Normalize normalizes param and sets default values
+func (param *QueryParameter) Normalize() {
+	param.Token = strings.ToUpper(param.Token)
+
+	if len(param.SQL) == 0 {
+		param.SQL = "?"
+	}
+}
+
+func (param *QueryParameter) ToValue(value interface{}) QueryParameterValue {
+	return QueryParameterValue{
+		Token: param.Type,
+		SQL:   param.SQL,
+		Value: value,
+	}
+}
+
+func (param *QueryParameter) ToDefaultValue() (QueryParameterValue, error) {
+	if len(param.DefaultValue) > 0 {
+		val, err := ParamTypeDecoders[param.Type](param.DefaultValue)
+		if err != nil {
+			return QueryParameterValue{}, err
+		}
+		return QueryParameterValue{
+			Token: param.Token,
+			SQL:   param.SQL,
+			Value: val,
+		}, nil
+	} else if len(param.DefaultSQL) > 0 {
+		return QueryParameterValue{
+			Token: param.Type,
+			SQL:   param.DefaultSQL,
+			Value: nil,
+		}, nil
+	} else {
```

Comment:
Hm... IMO we should support two cases:
- Parameter doesn't have `default_value` or `default_sql`, but it not marked as `required`. In that case we replace the token with an empty string.
- Like above, but the parameter is required. Then we can return this error (and make sure it ends up as 400).